Implement delete player service functionality

DarcoProgramador opened the pull request:
# Cambios

- Actualizando el puerto del servicio para eliminar un jugador
- Añadiendo funcionalidad al servicio para eliminar un jugador
- Añadiendo nuevos errores de dominio
- Actualizando el repositorio para eliminar un jugador y usar los errores de dominio

# Testting
![image](https://github.com/jairogloz/go-l/assets/110353857/adfd5ed6-b80a-4069-8a42-dcd77bcf7906)
No he subido aun el test porque aun no nos ponemos desacuerdo con como hacer los test, pero les dejo el código que utilice para los test en esta pr
``` go
package player

import (
	"fmt"
	"testing"

	"github.com/jairogloz/go-l/internal/domain"
)

type MockRepository struct {
	err error
}

func (r *MockRepository) Insert(player *domain.Player) (err error) {
	return nil
}

func (r *MockRepository) Get(id string) (player domain.Player, err error) {
	return domain.Player{}, nil
}

func (r *MockRepository) Delete(id string) (err error) {
	if r.err != nil {
		return r.err
	}
	return nil
}

func TestService_Delete(t *testing.T) {

	tests := []struct {
		name          string
		id            string
		repo          *MockRepository
		expectedError error
	}{
		{
			name: "Delete player successfully",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: nil,
			},
		},
		{
			name: "Delete player with incorrect id",
			id:   "id incorrecto",
			repo: &MockRepository{
				err: domain.ErrIncorrectID,
			},
			expectedError: domain.AppError{
				Code: domain.ErrCodeInvalidParams,
				Msg:  "error deleting player: incorrect id",
			},
		},
		{
			name: "Delete player not found",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: domain.ErrNotFound,
			},
			expectedError: domain.AppError{
				Code: domain.ErrCodeNotFound,
				Msg:  "error deleting player: player not found",
			},
		},
		{
			name: "Delete player with unknown error",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: fmt.Errorf("error deleting player"),
			},
			expectedError: domain.AppError{
				Code: domain.ErrCodeInternalServerError,
				Msg:  "error deleting player",
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			serv := Service{
				Repo: tt.repo,
			}

			err := serv.Delete(tt.id)
			if err != tt.expectedError {
				t.Errorf("expected error %v, got %v", tt.expectedError, err)
			}
		})
	}
}
```

DarcoProgramador committed:
Implement delete player service functionality

- Actualizando el puerto del servicio para eliminar un jugador
- Añadiendo funcionalidad al servicio para eliminar un jugador
- Añadiendo nuevos errores de dominio
- Actulizando el repositorio para eliminar un jugador para usar los errores de dominio

DarcoProgramador committed:
Implement delete player service functionality

- Actualizando el puerto del servicio para eliminar un jugador
- Añadiendo funcionalidad al servicio para eliminar un jugador
- Añadiendo nuevos errores de dominio
- Actulizando el repositorio para eliminar un jugador para usar los errores de dominio

jairogloz commented on internal/domain/app_error.go:
👏 excelente! que bien que seguiste el ejemplo de create player!

jairogloz commented on internal/services/player/delete.go:
```diff
@@ -1,8 +1,42 @@
+	return nil
+}
+
+func manageError(err error) error {
```
Buena propuesta! 👏 

Creo que podríamos hacer esta función más genérica y moverla a domain.app_error para que nos sirva no solo para delete palyer sino para todos los servicios.

Para esto también tenddríamos que hacer los mensajes más genéricos, por ejemplo, en lugar de "error deleting player: incorrect id" podría ser "incorrect id", y en lugar de "error deleting player: player not found" solo "not found". El punto es que los mensajes sean lo suficientemente genéricos para que nos sirvan para todos los servicios.

Luego podríamos parametrizar un poco y además del error recibir un `msg string` y si ese parámetro viene diferente de "" se puede concatenar con el mensaje genérico del error. 
```suggestion
func manageError(err error, msg string) error {
  var appErr domain.AppError
  switch err {
  case domain.ErrIncorrectID:
	log.Println("incorrect id error")
	appErr = domain.AppError{
		Code: domain.ErrCodeInvalidParams,
		Msg:  "error deleting player: incorrect id",
	}
	...
	}
	
	if msg != "" {
	  appErr.message := fmt.Sprintf("%s: appErr.message", msg)
	}
	return appErr
	
}
```

jairogloz reviewed (changes requested):
Excelente trabajo!!

Solo dejé una sugerencia para la función mangageError, tiene potencial para sernos muy útil!

Gracias!

DarcoProgramador commented:
hice un rabase a la rama para resolver el conflicto

DarcoProgramador committed:
add ManageError

- update delete services to implement ManageError
- add ManageError in domain package

DarcoProgramador commented:
> Excelente trabajo!!
> 
> Solo dejé una sugerencia para la función mangageError, tiene potencial para sernos muy útil!
> 
> Gracias!

Ya añadí la sugerencia de la función para manejar los errores de manera global

DarcoProgramador commented:
# Cambios 

- Se añadió una nueva función en el package de domain, para manejar los errores de servicios
- Se actualizo el servicio para eliminar un jugador para usar esta función global para manejar los errores

# Testing

Tuve que hacer unas modificaciones al código del testing para que usaran la función de manejar errores global del paquete de dominio
![image](https://github.com/jairogloz/go-l/assets/110353857/78a73515-b2ed-40da-bd8a-570bc7169666)

``` go
package player

import (
	"fmt"
	"testing"

	"github.com/jairogloz/go-l/internal/domain"
)

type MockRepository struct {
	err error
}

func (r *MockRepository) Insert(player *domain.Player) (err error) {
	return nil
}

func (r *MockRepository) Get(id string) (player domain.Player, err error) {
	return domain.Player{}, nil
}

func (r *MockRepository) Delete(id string) (err error) {
	if r.err != nil {
		return r.err
	}
	return nil
}

func TestService_Delete(t *testing.T) {

	tests := []struct {
		name          string
		id            string
		repo          *MockRepository
		expectedError error
	}{
		{
			name: "Delete player successfully",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: nil,
			},
		},
		{
			name: "Delete player with incorrect id",
			id:   "hola",
			repo: &MockRepository{
				err: domain.ErrIncorrectID,
			},
			expectedError: domain.ManageError(domain.ErrIncorrectID, "Error deleting player"),
		},
		{
			name: "Delete player not found",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: domain.ErrNotFound,
			},
			expectedError: domain.ManageError(domain.ErrNotFound, "Error deleting player"),
		},
		{
			name: "Delete player with unknown error",
			id:   "5f5b9d1e1c9d440000f9f3f5",
			repo: &MockRepository{
				err: fmt.Errorf("error deleting player"),
			},
			expectedError: domain.ManageError(fmt.Errorf("error deleting player"), "Error deleting player"),
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			serv := Service{
				Repo: tt.repo,
			}

			err := serv.Delete(tt.id)
			if err != tt.expectedError {
				t.Errorf("expected error %v, got %v", tt.expectedError, err)
			}
		})
	}
}
```

jairogloz committed:
Add custom message only if error code is not "internal server error"

jairogloz reviewed (approved):
Súper @DarcoProgramador !! Muchas gracias por los cambios 🚀
